src/Reflection/InitializerExprTypeResolver.php

```diff
@@ -1616,63 +1626,93 @@ private function integerRangeMath(Type $range, BinaryOp $node, Type $operand): T
 			} else {
-				$min = $rangeMin !== null && $operand->getMin() !== null && $operand->getMin() !== 0 ? $rangeMin / $operand->getMin() : null;
-				$max = $rangeMax !== null && $operand->getMax() !== null && $operand->getMax() !== 0 ? $rangeMax / $operand->getMax() : null;
-			}
+				// Avoid division by zero when looking for the min and the max by using the closest int
+				$operandMin = $operandMin !== 0 ? $operandMin : 1;
+				$operandMax = $operandMax !== 0 ? $operandMax : -1;
+
+				if (
+					($operandMin < 0 || $operandMin === null)
+					&& ($operandMax > 0 || $operandMax === null)
+				) {
+					$result = TypeCombinator::union(
+						$this->integerRangeMath($range, $node, IntegerRangeType::fromInterval($operandMin, 0)),
+						$this->integerRangeMath($range, $node, IntegerRangeType::fromInterval(0, $operandMax))
+					)->toNumber();
+
+					if ($result->equals(new UnionType([new IntegerType(), new FloatType()]))) {
+						return new BenevolentUnionType([new IntegerType(), new FloatType()]);
+					} else {
+						return $result;
+					}
+				}
+				if (
+					($rangeMin < 0 || $rangeMin === null)
+					&& ($rangeMax > 0 || $rangeMax === null)
+				) {
+					$result = TypeCombinator::union(
+						$this->integerRangeMath(IntegerRangeType::fromInterval($rangeMin, 0), $node, $operand),
+						$this->integerRangeMath(IntegerRangeType::fromInterval(0, $rangeMax), $node, $operand)
+					)->toNumber();
+
+					if ($result->equals(new UnionType([new IntegerType(), new FloatType()]))) {
+						return new BenevolentUnionType([new IntegerType(), new FloatType()]);
+					} else {
+						return $result;
+					}
+				}
+
+				$rangeMinSign = ($rangeMin ?? -INF) <=> 0;
+				$rangeMaxSign = ($rangeMax ?? INF) <=> 0;
+
+				$min1 = $operandMin !== null ? ($rangeMin ?? -INF) / $operandMin : $rangeMinSign * -0.1;
+				$min2 = $operandMax !== null ? ($rangeMin ?? -INF) / $operandMax : $rangeMinSign * 0.1;
+				$max1 = $operandMin !== null ? ($rangeMax ?? INF) / $operandMin : $rangeMaxSign * -0.1;
+				$max2 = $operandMax !== null ? ($rangeMax ?? INF) / $operandMax : $rangeMaxSign * 0.1;
+
+				$min = min($min1, $min2, $max1, $max2);
+				$max = max($min1, $min2, $max1, $max2);
 
-			if ($range instanceof IntegerRangeType && $operand instanceof IntegerRangeType) {
-				if ($rangeMax === null && $operand->getMax() === null) {
-					$min = 0;
-				} elseif ($rangeMin === null && $operand->getMin() === null) {
+				if ($min === -INF) {
 					$min = null;
+				}
+				if ($max === INF) {
 					$max = null;
 				}
 			}
 
+			if ($min !== null && $max !== null && $min > $max) {
+				[$min, $max] = [$max, $min];
+			}
+
 			if ($operand instanceof IntegerRangeType
-				&& ($operand->getMin() === null || $operand->getMax() === null)
 				|| ($rangeMin === null || $rangeMax === null)
-				|| is_float($min) || is_float($max)
+				|| is_float($min)
+				|| is_float($max)
 			) {
 				if (is_float($min)) {
-					$min = (int) $min;
+					$min = (int) ceil($min);
```

Comment:
`(int) 0.5` was 0.

So, `int<1, 2> / 2` was `int<0, 1>|float` which is wrong, 0 was never possible.

`(int) ceil(0.5)` is 1.
And `1|float` is the correct result.